From d3194da668eade182d4d463c5f064cd567a61738 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Tue, 26 Apr 2011 02:52:50 -0600
Subject: [PATCH 80/91] - added c2_cob_attr struct that will carry in-req
 information from reqh/mdservice down to mdstore; - cleanups in mdstore.

---
 cob/cob.h         |  22 ++++++++
 mdstore/mdstore.c | 147 ++++++++++++++++++++++++++++--------------------------
 2 files changed, 98 insertions(+), 71 deletions(-)

diff --git a/cob/cob.h b/cob/cob.h
index a6eb276..aad0e48 100644
--- a/cob/cob.h
+++ b/cob/cob.h
@@ -183,6 +183,28 @@ int c2_cob_domain_init(struct c2_cob_domain *dom, struct c2_dbenv *env,
                        struct c2_cob_domain_id *id);
 void c2_cob_domain_fini(struct c2_cob_domain *dom);
 
+/**
+   Attributes describing object that needs to be created or modified.
+   This structure is filled by mdservice and used in mdstore and/or
+   in cob modules for carrying in-request information to layers that
+   should not be dealing with fop req or rep.
+*/
+struct c2_cob_attr {
+        struct c2_fid     ca_fid;     /**< object fid */
+        int               ca_flags;   /**< marks valid fiedls. */
+        uint32_t          ca_mode;    /**< protection. */
+        uint32_t          ca_uid;     /**< user ID of owner. */
+        uint32_t          ca_gid;     /**< group ID of owner. */
+        uint64_t          ca_atime;   /**< time of last access. */
+        uint64_t          ca_mtime;   /**< time of last modification. */
+        uint64_t          ca_ctime;   /**< time of last status change. */
+        uint32_t          ca_nlink;   /**< number of hard links. */
+        uint32_t          ca_rdev;    /**< device ID (if special file). */
+        uint64_t          ca_size;    /**< total size, in bytes. */
+        uint64_t          ca_blksize; /**< blocksize for filesystem I/O. */
+        uint64_t          ca_blocks;  /**< number of blocks allocated. */
+        
+};
 
 /** 
    Namespace table. For data objects, pfid = cfid and name = ""
diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index 4b090ec..7aad167 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -36,10 +36,14 @@ struct c2_fid C2_MD_ROOT_FID = {
         .f_key       = 1ULL
 };
 
-int c2_md_store_init(struct c2_md_store *md, 
-                     struct c2_cob_domain_id *id,
-                     struct c2_dbenv *db, 
-                     int init_root)
+/**
+   Initialize mdstore on passed @id and db. Input argument @init_root
+   controls whether root cob should be initialized.
+*/
+int c2_md_store_init(struct c2_md_store         *md, 
+                     struct c2_cob_domain_id    *id,
+                     struct c2_dbenv            *db, 
+                     int                         init_root)
 {
         struct c2_db_tx        tx;
         int                    rc;
@@ -83,14 +87,14 @@ void c2_md_store_fini(struct c2_md_store *md)
 	c2_addb_ctx_fini(&md->md_addb);
 }
 
-int c2_md_store_create_attr(struct c2_md_store *md,
-                            struct c2_cob *pcob,
-                            struct c2_cob *cob,
-                            struct c2_cob_nskey *nskey,
-                            struct c2_cob_nsrec *nsrec,
+int c2_md_store_create_attr(struct c2_md_store   *md,
+                            struct c2_cob        *pcob,
+                            struct c2_cob        *cob,
+                            struct c2_cob_nskey  *nskey,
+                            struct c2_cob_nsrec  *nsrec,
                             struct c2_cob_fabrec *fabrec,
                             struct c2_cob_omgrec *omgrec,
-                            struct c2_db_tx *tx)
+                            struct c2_db_tx      *tx)
 {
         int rc;
         
@@ -141,12 +145,12 @@ int c2_md_store_create_attr(struct c2_md_store *md,
         return rc;
 }
 
-int c2_md_store_create(struct c2_md_store *md,
-                       struct c2_cob *pcob,
-                       struct c2_cob *cob,
-                       struct c2_fop_create *req,
+int c2_md_store_create(struct c2_md_store       *md,
+                       struct c2_cob            *pcob,
+                       struct c2_cob            *cob,
+                       struct c2_fop_create     *req,
                        struct c2_fop_create_rep *rep,
-                       struct c2_fop_ctx *ctx)
+                       struct c2_fop_ctx        *ctx)
 {
         struct c2_fop_cob     *body = &req->c_body;
         struct c2_cob_nskey   *nskey;
@@ -218,11 +222,11 @@ int c2_md_store_create(struct c2_md_store *md,
         return rc;
 }
 
-int c2_md_store_add_name(struct c2_cob *cob,
-                         struct c2_fid *pfid,
-                         const char *name,
-                         int namelen,
-                         struct c2_db_tx *tx)
+int c2_md_store_add_name(struct c2_cob          *cob,
+                         struct c2_fid          *pfid,
+                         const char             *name,
+                         int                     namelen,
+                         struct c2_db_tx        *tx)
 {
         struct c2_cob_nskey   *nskey;
         int                    rc;
@@ -234,12 +238,12 @@ int c2_md_store_add_name(struct c2_cob *cob,
         return rc;
 }
 
-int c2_md_store_link(struct c2_md_store *md, 
-                     struct c2_cob *pcob,
-                     struct c2_cob *cob, 
-                     struct c2_fop_link *req, 
-                     struct c2_fop_link_rep *rep, 
-                     struct c2_fop_ctx *ctx)
+int c2_md_store_link(struct c2_md_store         *md, 
+                     struct c2_cob              *pcob,
+                     struct c2_cob              *cob, 
+                     struct c2_fop_link         *req, 
+                     struct c2_fop_link_rep     *rep, 
+                     struct c2_fop_ctx          *ctx)
 {
         struct c2_cob_nskey   *nskey;
         time_t                 now;
@@ -285,12 +289,12 @@ out:
         return rc;
 }
 
-int c2_md_store_unlink(struct c2_md_store *md, 
-                       struct c2_cob *pcob,
-                       struct c2_cob *cob,
-                       struct c2_fop_unlink *req, 
+int c2_md_store_unlink(struct c2_md_store       *md,
+                       struct c2_cob            *pcob,
+                       struct c2_cob            *cob,
+                       struct c2_fop_unlink     *req, 
                        struct c2_fop_unlink_rep *rep, 
-                       struct c2_fop_ctx *ctx)
+                       struct c2_fop_ctx        *ctx)
 {
         struct c2_cob         *ncob;
         struct c2_cob_nskey   *nskey;
@@ -408,11 +412,11 @@ out:
         return rc;
 }
 
-int c2_md_store_open(struct c2_md_store *md, 
-                     struct c2_cob *cob,
-                     struct c2_fop_open *req, 
-                     struct c2_fop_open_rep *rep, 
-                     struct c2_fop_ctx *ctx)
+int c2_md_store_open(struct c2_md_store         *md, 
+                     struct c2_cob              *cob,
+                     struct c2_fop_open         *req, 
+                     struct c2_fop_open_rep     *rep, 
+                     struct c2_fop_ctx          *ctx)
 {
         int rc = 0;
         
@@ -448,9 +452,10 @@ int c2_md_store_close(struct c2_md_store *md,
         return rc;
 }
 
-static int c2_md_store_rename_sanity(struct c2_md_store *md, 
-                                     struct c2_fop_rename *rename,
-                                     struct c2_fop_ctx *ctx)
+static int 
+c2_md_store_rename_sanity(struct c2_md_store    *md, 
+                          struct c2_fop_rename  *rename,
+                          struct c2_fop_ctx     *ctx)
 {
         struct c2_cob_oikey  oikey;
         struct c2_fid        tgtfid;
@@ -500,13 +505,13 @@ static int c2_md_store_rename_sanity(struct c2_md_store *md,
         return rc;
 }
 
-int c2_md_store_rename(struct c2_md_store *md, 
-                       struct c2_cob *pcob_tgt,
-                       struct c2_cob *pcob_src,
-                       struct c2_cob *cob,
-                       struct c2_fop_rename *req, 
+int c2_md_store_rename(struct c2_md_store       *md, 
+                       struct c2_cob            *pcob_tgt,
+                       struct c2_cob            *pcob_src,
+                       struct c2_cob            *cob,
+                       struct c2_fop_rename     *req, 
                        struct c2_fop_rename_rep *rep, 
-                       struct c2_fop_ctx *ctx)
+                       struct c2_fop_ctx        *ctx)
 {
         struct c2_cob_nskey  *srckey;
         struct c2_cob_nskey  *tgtkey;
@@ -612,11 +617,11 @@ out:
         return rc;
 }
 
-int c2_md_store_setattr(struct c2_md_store *md, 
-                        struct c2_cob *cob,
-                        struct c2_fop_setattr *req, 
+int c2_md_store_setattr(struct c2_md_store      *md,
+                        struct c2_cob           *cob,
+                        struct c2_fop_setattr   *req, 
                         struct c2_fop_setattr_rep *rep, 
-                        struct c2_fop_ctx *ctx)
+                        struct c2_fop_ctx       *ctx)
 {
         struct c2_fop_cob     *body = &req->s_body;
         struct c2_cob_nsrec   *nsrec;
@@ -669,11 +674,11 @@ int c2_md_store_setattr(struct c2_md_store *md,
         return rc;
 }
 
-int c2_md_store_getattr(struct c2_md_store *md, 
-                        struct c2_cob *cob,
-                        struct c2_fop_getattr *req, 
+int c2_md_store_getattr(struct c2_md_store      *md, 
+                        struct c2_cob           *cob,
+                        struct c2_fop_getattr   *req, 
                         struct c2_fop_getattr_rep *rep, 
-                        struct c2_fop_ctx *ctx)
+                        struct c2_fop_ctx       *ctx)
 {
         struct c2_fop_cob *body = &rep->g_body;
         int                rc = 0;
@@ -718,11 +723,11 @@ int c2_md_store_getattr(struct c2_md_store *md,
         return rc;
 }
 
-int c2_md_store_readdir(struct c2_md_store *md, 
-                        struct c2_cob *cob,
-                        struct c2_fop_readdir *req, 
+int c2_md_store_readdir(struct c2_md_store      *md, 
+                        struct c2_cob           *cob,
+                        struct c2_fop_readdir   *req, 
                         struct c2_fop_readdir_rep *rep, 
-                        struct c2_fop_ctx *ctx)
+                        struct c2_fop_ctx       *ctx)
 {
         struct c2_cob_iterator         it;
         struct c2_fop_readdir_vec     *vec;
@@ -781,11 +786,11 @@ out:
 /**
    Find cob by fid.
 */
-int c2_md_store_locate(struct c2_md_store *md, 
-                       struct c2_fid *fid,
-                       struct c2_cob **cob, 
-                       int flags, 
-                       struct c2_db_tx *tx)
+int c2_md_store_locate(struct c2_md_store       *md, 
+                       struct c2_fid            *fid,
+                       struct c2_cob           **cob, 
+                       int                       flags, 
+                       struct c2_db_tx          *tx)
 {
         struct c2_cob_oikey oikey;
         int                 rc;
@@ -807,12 +812,12 @@ int c2_md_store_locate(struct c2_md_store *md,
 /**
    Find cob by parent and name.
 */
-int c2_md_store_lookup(struct c2_md_store *md, 
-                       struct c2_cob *pcob,
-                       const char *name, 
-                       int namelen, 
-                       struct c2_cob **cob, 
-                       struct c2_db_tx *tx)
+int c2_md_store_lookup(struct c2_md_store       *md, 
+                       struct c2_cob            *pcob,
+                       const char               *name, 
+                       int                       namelen, 
+                       struct c2_cob           **cob, 
+                       struct c2_db_tx          *tx)
 {
         struct c2_cob_nskey *nskey;
         int                  flags;
@@ -831,9 +836,9 @@ int c2_md_store_lookup(struct c2_md_store *md,
 /**
    Allocate new cob on mdstore @md.
 */
-int c2_md_store_alloc(struct c2_md_store *md, 
-                      struct c2_fid *fid, 
-                      struct c2_cob **cob)
+int c2_md_store_alloc(struct c2_md_store        *md, 
+                      struct c2_fid             *fid, 
+                      struct c2_cob            **cob)
 {
         int  rc;
         
-- 
1.8.3.2

